cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -592,7 +603,7 @@ func (mgr *packetManagerRest) getNodes(nodegroup string) ([]string, error) {
 
 	for _, d := range devices.Devices {
 		if Contains(d.Tags, "k8s-cluster-"+mgr.getNodePoolDefinition(nodegroup).clusterName) && Contains(d.Tags, "k8s-nodepool-"+nodegroup) {
-			nodes = append(nodes, fmt.Sprintf("packet://%s", d.ID))
+			nodes = append(nodes, fmt.Sprintf("equinixmetal://%s", d.ID))
```

Comment:
If I'm not mistaken, wouldn't this cause the autoscaler to think that any Nodes w/ the 'packet://' provider id prefix to be unregistered (and as a result, eventually deleted)?

https://github.com/kubernetes/autoscaler/blob/c64e97570a2f06dabe685e6f0387495431466217/cluster-autoscaler/clusterstate/clusterstate.go#L295-L329

I've hit similar complications around the Node provider IDs with cluster-api-provider-packet around migration from an older version of the packet-ccm to a newer version of cloud-provider-equinix-metal, and the migration process can be a bit messy, especially since the ProviderID field on the Node is immutable.

There are probably a couple of options available:
- Switch everything to use the new `equinixmetal:;//` provider id prefix and require using a newer version of the cloud-provider
- Make the prefix configurable, so that it can work with either prefix, but the user must configure it correctly

Either way, we should probably document some type of migration strategy, since I'm not sure there is a way to stay flexible without changes to the core autoscaling code, or forcing node/device churn in the cluster.

For cluster-api-provider-packet, the current plan is to do something along the lines of:
- Update packet-ccm to cloud-provider-equinix-metal
- Upgrade the cluster api provider to a version that uses the new provider id prefix
- Delete and Recreate the Node resources to change the providerID.

A similar approach should be possible with the autoscaler, thought it might be better to disable the autoscaler during the upgrade process to avoid any potential of confusion that could cause workload disruption during the node deletion/recreation process.

Reply:
Good point @detiber. Thanks for the feedback :)

@ipochi as far as I understand, it is not possible to get mixed `packet://` and `equnixmetal://` on a single cluster, right? Those are set by the CCM, which will either set it to one or another, or? If that's the case, we should indeed make it configurable.

Reply:
@detiber Thanks for your valuable feedback. I think I misunderstood about the importance of `packet://` in
```nodes = append(nodes, fmt.Sprintf("packet://%s", d.ID))
```

Also @invidian @detiber 

Its is quite possible to get mixed `packet://` and `equinixmetal://` on a single cluster. If ta cluster is created with packet-ccm and then packet-ccm is upgraded to cloud-provider-equinix-metal, then any new nodes created will have `equinixmetal://`, whilst the old nodes would still continue to have `packet://`

Example:
```
$ kubectl get nodes
NAME                         STATUS   ROLES    AGE     VERSION
theta-anotherpool-worker-0   Ready    <none>   9m29s   v1.21.1 - After upgrade to cloud-provider-equinix-metal
theta-controller-0           Ready    <none>   86m     v1.21.1 - Before upgrade
theta-small-worker-0         Ready    <none>   86m     v1.21.1 - Before upgrade
```

```
$ kubectl get nodes -o jsonpath='{.items[*].spec.providerID}'
equinixmetal://536e117a-6d6e-4d9c-b502-f67170662aef packet://b7057a83-732e-4ae6-b5a7-d28516556ba6 packet://4815edc6-b0e2-4bdb-bbb0-daca73e72c1a⏎
```

Reply:
>Its is quite possible to get mixed packet:// and equinixmetal:// on a single cluster. If ta cluster is created with packet-ccm and then packet-ccm is upgraded to cloud-provider-equinix-metal, then any new nodes created will have equinixmetal://, whilst the old nodes would still continue to have packet://

Yes, but this is the situation which can be gated. You can disable the autoscaler, do CCM -> CCEM upgrade, re-register all nodes, then upgrade the autoscaler I guess.

Reply:
The scenario I described above didn't involve autoscaler.

Merely upgrading from CCM TO CCEM and adding a node to the cluster.


Can deregistering and  registering a node be done in case of a single controller node ?

Reply:
>The scenario I described above didn't involve autoscaler.

> Merely upgrading from CCM TO CCEM and adding a node to the cluster.

It doesn't matter who adds new node :) But the autoscaler can do it unattended. 

>Can deregistering and registering a node be done in case of a single controller node ?

That seem Lokomotive specific question and I don't know the answer :smile:

Reply:
Q: We should't worry about controller nodes if I understand correctly ?

Reply:
Assuming they will be correctly skipped, yes. Good point!